reps/2023-10-30-gpu-memory-support.md

```diff
@@ -0,0 +1,208 @@
+
+Inside `ray` project since we want to add new parameter `gpu_memory` to the `ray` remote function.
+
+## Stewardship
+
+### Required Reviewers
+The proposal will be open to the public, but please suggest a few experienced Ray contributors in this technical domain whose comments will help this proposal. Ideally, the list should include Ray committers.
+
+@pcmoritz, @jjyao, @scv119
+
+### Shepherd of the Proposal (should be a senior committer)
+To make the review process more productive, the owner of each proposal should identify a **shepherd** (should be a senior Ray committer). The shepherd is responsible for working with the owner and making sure the proposal is in good shape (with necessary information) before marking it as ready for broader review.
+
+@jjyao
+
+## Design
+
+### API
+Users will be able to specify amount of GPU memory to their ray tasks/actors using `gpu_memory` on `ray.remote`. The specified `gpu_memory` will be the amount of GPU resources from a single GPU that will be allocated to users ray tasks/actors.
+
+```python
+# Request a fractional GPU with specified gpu_memory in bytes.
+# Mutually exclusive with num_gpus.
+@ray.remote(gpu_memory=1024 * 1024 * 1024) # 1 mb request
+def task:
+  …
+```
+
+When a Ray node is started, Ray will auto detect the number of GPUs and GPU memory of each GPU. Users also have the option to manually specify them:
+
+```bash
+ray start # auto detection
+
+ray start –num_gpus=3 –gpu_memory=1000 * 1024 * 1024 * 1024 # manual override, each GPU has 1000mb memory
+```
+
+Note that GPU memory and GPU is 1-1 conversion, means 20GB of GPU memory is equivalent to 0.5 fractional value of an `A100_40GB` GPU. So, for simplicity and consistency, ray doesn't allow users to specify both `num_gpus` and `gpu_memory` in a single ray task/actor.
+
+
+```python
+# Request a fractional GPU both num_gpus and gpu_memory is not allowed
+@ray.remote(gpu_memory=1024 * 1024 * 1024, num_gpus=0.5) # raise ValueError exception
+def not_allowed_task:
+  …
+
+# Request a fractional GPU with specified num_gpus.
+@ray.remote(num_gpus=0.5)
+def num_gpus_task:
+  …
+
+# Request a fractional GPU with specified gpu_memory.
+@ray.remote(gpu_memory=1024 * 1024 * 1024) 
+def gpu_memory_task:
+  …
+```
+
+Additionally, users can still specify which GPU type they want to use by specifying `accelerator_type`.
+
+```python
+# Request a fractional of A100 GPU with specified gpu_memory
+@ray.remote(gpu_memory=1024 * 1024 * 1024 * 1024, accelerator_type="NVIDIA_A100")
+def nvidia_a100_gpu_task:
+  …
+
+# Requesting 30GB of GPU memory from a A10 GPU with 24GB of memory.
+# Task won't be able to be scheduled.
+@ray.remote(gpu_memory=30 * 1024 * 1024 * 1024 * 1024, accelerator_type="NVIDIA_TESLA_A10G")
+def nvidia_a10_gpu_task:
+  …
+```
+
+#### Placement Group
+User is also able to request `gpu_memory` in placement group bundles as follows:
+
+```python
+pg = placement_group([{"gpu_memory": 1024 * 1024, "CPU": 1}, {"GPU": 1}])
+```
+
```

Comment:
I think we need the observability section here as this complicates the observability semantics. 

- how is it displayed in ray status?
    - for ray status, it should potentially display sth like gpu_memory: 4 gpus (A10) * 3gb? 
    - In ray status, if a task is scheduled with gpu_memory, both gpu & gpu memory values are subtracted? 
- How is it displayed in resource_requirement in ray list tasks? Is it translated into num_gpus? Or it only includes gpu_memory? Or both?

Reply:
in `ray list nodes`, it will be `GPU (resources left) * gpu_memory_per_gpu` which is the constant stored in node label. `ray status`, `ray list task` and `ray.available_resources` currently didn't show GPU memory but if we added one, it will be the same as `ray list nodes`.

and yes, basically both gpu and gpu_memory values are subtracted to show the remaining